src/editor.rs

```diff
@@ -559,56 +454,48 @@ impl<'arena, Node: Ast<'arena> + 'arena> Editor<'arena, Node> {
     }
 
     /// Consumes a [`char`] and adds it to the command buffer.  If the command buffer contains a
-    /// valid command, then execute that command.  This returns `true` if the command 'Quit' was
-    /// executed, otherwise `false` is returned.
-    fn consume_command_char(&mut self, c: char) -> bool {
-        let mut should_quit = false;
+    /// valid command, then execute that command.
+    ///
+    /// This returns a tuple of:
+    /// 1. A [`bool`] value that determines whether or not Sapling should quit
+    /// 2. The [`EditResult`] of the edit, or `None` if the command is incomplete
+    fn consume_command_char(&mut self, c: char) -> (bool, Option<EditResult>) {
         // Add the new keypress to the command
         self.command.push(c);
         // Attempt to parse the command, and take action if the command is
         // complete
-        if let Some(action) = parse_command(&self.keymap, &self.command) {
-            // Respond to the action
-            match action {
-                Action::Undefined => {
-                    log::warn!("'{}' is not a command.", self.command);
-                }
-                Action::Quit => {
-                    // Break the mainloop to quit
-                    log::trace!("Received command 'Quit', so exiting mainloop");
-                    should_quit = true;
-                }
-                Action::MoveCursor(direction) => {
-                    self.move_cursor(direction);
-                }
-                Action::Replace(c) => {
-                    self.replace_cursor(c);
-                }
-                Action::InsertChild(c) => {
-                    self.insert_child(c);
-                }
-                Action::InsertBefore(c) => {
-                    self.insert_next_to_cursor(c, Side::Prev);
-                }
-                Action::InsertAfter(c) => {
-                    self.insert_next_to_cursor(c, Side::Next);
-                }
-                Action::Delete => {
-                    self.delete_cursor();
-                }
-                Action::Undo => {
-                    self.undo();
-                }
-                Action::Redo => {
-                    self.redo();
-                }
+        match parse_command(&self.keymap, &self.command) {
+            Some(action) => {
+                let mut should_quit = false;
+                // Respond to the action
+                let result = match action {
```

Comment:
Nested `match` can be sometimes confusing and I like to avoid it. Either by matching like this:

```rust
let mut should_quit = false;
match parse_command(&self.keymap, &self.command) {
    Some(Action::Undefined) => Err(EditErr::Invalid(self.command.clone())),
    Some(Action::Undo) => self.tree.undo(),
    // ...
    None => (false, None),
}
```

Or by short circuiting the outer match.

Reply:
I've refactored the inner `match` into a separate function, which just feels better anyway.